src/iop/expfusion.c

```diff
@@ -0,0 +1,1268 @@
+}
+
+static inline void _images_div(const float *const img_src1, const size_t wd, const size_t ht, const int ch,
+                               const float *const img_src2, float *const img_dest)
+{
+  const size_t size = wd * ht * ch;
+
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(size, img_src1, img_src2, img_dest) \
+  schedule(static)
+#endif
+  for(int i = 0; i < size; i++)
+  {
+    if(img_src2[i] != 0.f) img_dest[i] = img_src1[i] / img_src2[i];
+  }
+}
+
+static inline void _images_add(const float *const img_src1, const size_t wd, const size_t ht, const int ch,
+                               const float *const img_src2, float *const img_dest)
+{
+  const size_t size = wd * ht * ch;
+
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(size, img_src1, img_src2, img_dest) \
+  schedule(static)
+#endif
+  for(int i = 0; i < size; i++)
+    img_dest[i] = img_src1[i] + img_src2[i];
+}
+
+static inline void _images_sub(const float *const img_src1, const size_t wd, const size_t ht, const int ch,
+                                   const float *const img_src2, float *const img_dest)
+{
+  const size_t size = wd * ht * ch;
+
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(size, img_src1, img_src2, img_dest) \
+  schedule(static)
+#endif
+  for(int i = 0; i < size; i++)
+    img_dest[i] = img_src1[i] - img_src2[i];
+}
+
+static inline void _image_add(const float *const img_src, const size_t wd, const size_t ht, const int ch,
+                              const float val, float *const img_dest)
+{
+  const size_t size = wd * ht * ch;
+
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(size, val, img_src, img_dest) \
+  schedule(static)
+#endif
+  for(int i = 0; i < size; i++)
+    img_dest[i] = img_src[i] + val;
+}
+
+static void _convolve_symmetric(const float *const img_src, const size_t wd, const size_t ht, const int ch,
+                                const float *const fx, const float *const fy, float *const img_tmp,
+                                float *const img_dest)
+{
+  // horizontal filter
+  for(int i = 0; i < ht; i++) // all lines
+  {
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(img_src, img_tmp, fx, wd, ch, i) \
+  schedule(static)
+#endif
+    for(int j = 2; j < wd - 2; j++)
+      for(int k = 0; k < ch; k++)
+        img_tmp[(i * wd + j) * ch + k]
+            = img_src[((i)*wd + (j - 2)) * ch + k] * fx[0] + img_src[((i)*wd + (j - 1)) * ch + k] * fx[1]
+              + img_src[((i)*wd + (j)) * ch + k] * fx[2] + img_src[((i)*wd + (j + 1)) * ch + k] * fx[3]
+              + img_src[((i)*wd + (j + 2)) * ch + k] * fx[4];
+
+    // left edge
+    int j = 0; // 1 0 [0 1 2 ... ]
+    for(int k = 0; k < ch; k++)
+      img_tmp[(i * wd + j) * ch + k]
+          = img_src[((i)*wd + (j + 1)) * ch + k] * fx[0] + img_src[((i)*wd + (j)) * ch + k] * fx[1]
```

Comment:
It looks like you're mirroring here - have you confirmed that this is indeed what enfuse is doing in their expand() function as opposed to the extrapolation described in Burt & Adelson 1983 (which Mertens derives their method from)?

Reply:
I didn't look into enfuse that much, my goal is not to have the exact same result as enfuse but a good/consistent result. If enfuse give better results then we can check it.

Reply:
I'm fine with that.  Enfuse may not use B&A's approach either - but I think their extrapolation approach may potentially provide better results.

Instead of mirroring across a boundary, or sample-and-hold, they do (hopefully my notation isn't too unclear) for the left/top boundary:
g(-l) = 2*g(0) - g(l)
As opposed to using g(0) for sample-and-old or g(l) for mirroring

I guess you could describe it as inverted mirroring?

Reply:
If I'm not mistaken this is the author's approach, but that doesn't mean that it cannot be improven. But I don't see why inverted mirror can be better.

Reply:
At least the theory on inverted mirror is that the first derivative in the spatial domain remains constant across the boundary, and the second derivative is zero.

I need to dig into mertens' Matlab code to see how they handled boundary conditions.

It appears that Enfuse is performing some sort of scaling of the pixel values when expanding, but it's not entirely clear what their intent is.  They do strongly hint that they are doing extrapolation (which is probably a better term to use than "inverse mirror").

Moving from mirroring to extrapolation seemed to offer a slight improvement when I tried it, I'll take a look tonight and see if it's beneficial in your implementation.  Since you've unrolled all of the for() loops, it will be easier to tweak the boundary condition handling.  :)